### PR TITLE
Python mms

### PR DESCRIPTION
**Pull request by aeslaughter**

### CODE REVIEW

**Comment by aeslaughter** [comment]
@lindsayad https://mooseframework.inl.gov/docs/PRs/13190/site/python/mms.html

File: modules/doc/content/python/mms.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+\begin{equation}
+\label{strong}
+\nabla \cdot \nabla u = 0.
+\end{equation}
+
+Next, assumed a solution to this equation. It is important to pick an equation that has the
+necessary order to allow for all the derivatives to be computed. It is also important that the
+finite element shape functions can not represent the solution exactly. Functions that
+include trigonometric or exponential terms are often an adequate choice. For this example the assumed
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply modules/doc/content/python/mms.md]: Of course if you're doing a steady simulation then you want the true solution to be higher order than the approximation polynomials or else you'll just get an L2 error of zero (although that's a check in and of itself).

File: modules/doc/content/python/mms.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+For this problem the error is computed using the [ElementL2Error.md] postprocessor and the
+element size is computed using the [AverageElementSize.md] postprocessor. If the mesh is not
+uniform it may be more applicable to use the number of degrees of freedom as a proxy for element
+size, which can be computed using the [NumDOFs.md] postprocessor.
+
+## Convergence Plot
+
+To determine if the convergence is correct the results from the above input file may be analyzed
+by plotting the data on a log-log plot. The 'mms' package includes a 'plot' function designed for
+this application that accepts the of the outputed CSV file(s). For this problem the following
```
Comment (lindsayad) [inline modules/doc/content/python/mms.md]: "accepts the of"?

`outputed` -> `outputted`

**Review by lindsayad — DISMISSED** [review]
I like the direction of this PR. I'm wondering whether we could automate things even more? For instance for the transient mms I just did, I ran something like:
```python
results = np.zeros(len(dts))
index,dt in enumerate(dts):
args = ['../../../../navier_stokes-opt',
        '--allow-test-objects',
       '-i',
       'stabilized-time-convergence-study.i',
       'Functions/forcing_function/value=' + str(forcing_function),
       'Functions/exact_soln/value=' + str(exact_soln),
       'Executioner/dt=' + str(dt),
       'Executioner/TimeIntegrator/type=' + time_integrator,
       'GlobalParams/tau_type=' + tau_scheme,
       "Kernels/active=" + supg.active_kernels,
       "Kernels/supg/include_transient_term=" + supg.non_time_flag]
subprocess.run(args)
results[index] = get_last_row('stabilized-time-convergence-study_out.csv')[1]
```
Or perhaps it's not too much of a stretch to have this just be user code. If the latter, then it would be really great to have a README or something in your mms test directory that described exactly how you generated the `.csv` files that you used for plotting...and ideally that generation is done automatically. Or even better than a README...a python script that does the whole kit and kaboodle from generating the forcing functions to calling your plot functions.

**Comment by aeslaughter** [comment]
@lindsayad Let me think of how to wrap up the execution, just leave this open for now.

**Comment by rwcarlsen** [comment]
I agree with alex - lets either automate a bit more or be sure we script and/or document the full sequence used for generating mms tests.

**Comment by aeslaughter** [comment]
@rwcarlsen @lindsayad The doco is updated https://mooseframework.inl.gov/docs/PRs/13190/site/python/mms.html

**Comment by aeslaughter** [comment]
This PR is reporting a conflict and I can't figure out why, so I am killing it.

File: modules/doc/content/python/mms.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+\label{strong}
+\nabla \cdot \nabla u = 0.
+\end{equation}
+
+Next, assume a solution to this equation. It is important to pick an equation that has the necessary
+order to allow for all the derivatives to be computed, both spatially and temporally.
+
+When performing a spatial convergence study for a transient problem the time portion
+should be an appropriate order that allows for the temporal finite difference scheme to represent
+solution exactly. To the contrary the spatial solution should not be exactly represented
```
Comment (lindsayad) [inline modules/doc/content/python/mms.md]: Did you find that you could do this? That you were able to eliminate possible error pollution from the temporal time integration?

**Review by lindsayad — APPROVED** [review]
I think the addition of the `run_spatial` and `run_temporal` functions is great. I like the `kwargs` for passing command line options. I think I would use the latter feature in conjunction with the forcing function generation functions within the same python script so that I could easily change my exact solution and have it updated in my input file using the command line arguments.

The next time I need to use mms for my own work, I may add to the documentation here demonstrating the use I discussed above. You've done enough here for this to go in IMO

**Review by rwcarlsen — APPROVED** [review]
It seems less bad enough I suppose.

File: modules/doc/content/python/mms.md (hunk 3 above)
Comment (aeslaughter) [reply modules/doc/content/python/mms.md]: I think I did, I did this in the results from the two examples and both achieved the convergence expected.

**Pull request merged**